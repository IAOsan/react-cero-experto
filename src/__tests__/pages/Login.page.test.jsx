import { vi } from 'vitest';
import '@testing-library/jest-dom';
import { MemoryRouter, Routes, Route } from 'react-router-dom';
import { renderWithProviders, setupUser } from '../test-utils';
import Login from '../../pages/Login.page';
import {
	loginWithGoogle,
	loginWithEmailPassWord,
} from '../../store/slices/auth/auth.slice';

vi.mock('../../store/slices/auth/auth.slice', async () => {
	const org = await vi.importActual('../../store/slices/auth/auth.slice');

	return {
		...org,
		loginWithGoogle: vi.fn(),
		loginWithEmailPassWord: vi.fn(() => vi.fn()),
	};
});

const user = setupUser();

describe('<Login > page', () => {
	beforeEach(() => {
		vi.clearAllMocks();
	});

	it('should renders correctly', () => {
		const wrapper = renderLogin();
		// snapshot
		expect(wrapper.container.firstElementChild).toMatchSnapshot();
		// title
		expect(wrapper.getByText(/log in/i)).toBeInTheDocument();
		// login with social
		expect(
			wrapper.getByText(/Login with social networks/i)
		).toBeInTheDocument();
		// google btn
		expect(
			wrapper.getByRole('button', { name: /sign in with google/i })
		).toBeInTheDocument();
		// footer to register
		expect(wrapper.getByText(/Need an account?/i)).toBeInTheDocument();
		expect(
			wrapper.getByRole('link', { name: /register/i })
		).toBeInTheDocument();
	});
	it('should be able to login with email and password', async () => {
		const userData = {
			email: 'test@mail.com',
			password: '123456',
		};
		const wrapper = renderLogin();

		await user.type(wrapper.getByPlaceholderText(/email/i), userData.email);
		await user.type(
			wrapper.getByPlaceholderText(/password/i),
			userData.password
		);
		await user.keyboard('{Enter>1}');

		expect(loginWithEmailPassWord).toHaveBeenCalledWith(userData);
	});
	it('should be able to login with google', async () => {
		const wrapper = renderLogin();

		await user.click(wrapper.getByText(/Sign in with Google/i));

		expect(loginWithGoogle).toHaveBeenCalled();
	});
	it('should redirects to register page when clicks link', async () => {
		const wrapper = renderLogin();

		await user.click(wrapper.getByRole('link', { name: /REGISTER/i }));

		expect(wrapper.queryByText(/log in/i)).toBeNull();
		expect(wrapper.queryByRole('button', { name: /login/i })).toBeNull();
	});
});

function renderLogin(entries = ['/auth/login']) {
	return renderWithProviders(
		<MemoryRouter initialEntries={entries}>
			<Routes>
				<Route path='/auth/login' element={<Login />} />
				<Route path='/auth/register' element={<mock-register />} />
			</Routes>
		</MemoryRouter>
	);
}
