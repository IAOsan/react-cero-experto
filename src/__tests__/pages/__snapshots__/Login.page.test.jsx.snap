// Vitest Snapshot v1

exports[`<Login > page > should renders correctly 1`] = `
<form
  class="form anima-fade-in-bck"
>
  <h2
    class="heading-2 mb-24 text-primary-alt"
  >
    Log in
  </h2>
  <div
    class="form__group"
  >
    <input
      class="form__control"
      id="inputEmail"
      name="email"
      placeholder="Email"
      type="text"
      value=""
    />
  </div>
  <div
    class="form__group"
  >
    <input
      class="form__control"
      id="inputPassword"
      name="password"
      placeholder="Password"
      type="password"
      value=""
    />
  </div>
  <button
    class="btn btn--default btn--primary btn--block my-20"
    disabled=""
    type="submit"
  >
    <div
      class="flex flex-ai-c flex-jc-c"
    >
      Login
    </div>
  </button>
  <hr />
  <p
    class="text-center text-muted my-12"
  >
    <strong>
      Login with social networks
    </strong>
  </p>
  <button
    class="btn google-btn"
    type="button"
  >
    <div
      class="google-icon-wrapper"
    >
      <img
        alt="google button"
        class="google-icon"
        src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
      />
    </div>
    <p
      class="btn-text"
    >
      Sign in with Google
    </p>
  </button>
  <p
    class="mt-20 text-center"
  >
    Need an account?
    <a
      class="mx-8 text-underline"
      href="/auth/register"
    >
      REGISTER
    </a>
  </p>
</form>
`;
